manual/tools/caml_tex2.ml

```diff
@@ -248,6 +294,85 @@ type incompatibility =
   | Signature_with_visible_answer of string * int
 exception Incompatible_options of incompatibility
 
+exception Phrase_parsing of string
+
+module Ellipsis = struct
+  (** This module implements the extraction of ellipsis locations
+      from phrases.
+
+      An ellipsis is either an [[@ellipsis]] attribute, or a pair
+      of [[@@@ellipsis.start]...[@@@ellipsis.stop]] attributes. *)
+
+  exception Unmatched_ellipsis of {kind:string; start:int; stop:int}
+  (** raised when an [[@@@ellipsis.start]] or [[@@@ellipsis.stop]] is
+      not paired with another ellipsis attribute *)
+
+  exception Nested_ellipses of {first:int ; second:int }
+  (** raised by [[@@@ellipsis.start][@@@ellipsis.start]] *)
+
+  let extract f x =
+    let transforms = ref [] in
+    let last_loc = ref Location.none in
+    let left_mark = ref None (* stored position of [@@@ellipsis.start]*) in
+    let location _this loc =
+      (* we rely on the fact that the default iterator call first
+         the location subiterator, then the attribute subiterator *)
+      last_loc := loc in
+    let attribute _this (attr,_) =
+      let name = attr.Location.txt in
+      let loc = !last_loc in
+      let start = loc.Location.loc_start.Lexing.pos_cnum in
+      let attr_start = attr.Location.loc.loc_start.Lexing.pos_cnum in
+      let attr_stop = 1 + attr.Location.loc.loc_end.Lexing.pos_cnum in
+      let stop = loc.Location.loc_end.Lexing.pos_cnum
+      in
+      match name with
+      | "ellipsis" ->
+          transforms :=
+            { Text_transform.kind=Ellipsis; start; stop= max attr_stop stop }
+            :: !transforms
```

Comment:
Shouldn't you check that `left_mark` is not active to guarantee the absence of nested ellipses?

Reply:
Possibly. However, this case is already caught later by `check_partition`. I will think a bit about this independence problem.

Reply:
I have added a test in order to catch nested ellipsis early in all cases (which makes the the test for unmatched stop/start simpler).